### ISSUE TITLE
StringTransport should behave more like a real transport

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1844]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)| @washort reported|
|-|-|
|Trac ID|trac#1844|
|Type|enhancement|
|Created|2006-06-22 14:20:49Z|

in particular, StringTransport.write should raise a TypeError when unicode is passed to it.

<details><summary>Searchable metadata</summary>

```
trac-id__1844 1844
type__enhancement enhancement
reporter__washort washort
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1150986049000000 1150986049000000
changetime__1177314977000000 1177314977000000
version__None None
owner__ 
cc__dash cc__exarkun cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort set owner to @exarkun|
|-|-|

Please review in stringtransport-fix-1844.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @washort|
|-|-|

Excellent.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Oops, actually: not quite.  trial won't pick up unit tests in a file named proto_helpers.py.  Move the test case into another module?  Then it should be good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Ping?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Moved in test_protocols.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Merged forward to stringtransport-1844-2 to resolve conflicts.  Looks good now.  Take a look at the new diff and merge if you think I didn't introduce anything crazy in the new branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [20040]) Merge stringtransport-fix-1844-2

Authors: washort, therve
Reviewer: exarkun
Fixes [#1844](https://github.com/twisted/twisted/issues/1844)

Fix StringTransport from proto_helpers: raises a TypeError when unicode is
passed, so that it behaves more like a real transport.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
